src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -550,12 +550,13 @@ def _validate_data_node(dim_node):
                 for dim in dimension:
                     if isinstance(dim, DataNode):
                         _validate_data_node(dim)
-                    elif not isinstance(dim, int):
+                    elif not isinstance(dim, (int, ArrayType.Extent)):
                         raise TypeError(
                             f"A DataSymbol shape-list element specifying lower"
                             f" and upper bounds must be a 2-tuple containing "
-                            f"either int or DataNode entries but '{dimension}'"
-                            f" contains '{type(dim).__name__}'")
+                            f"either int, DataNode or ArrayType.Extent entries"
+                            f" but '{dimension}' contains "
+                            f"'{type(dim).__name__}'")
```

Comment:
Could we also validate that the upper can not have a value if the lower doesn't? For instance now this is accepted:
` array_type = ArrayType(INTEGER_TYPE, [(ArrayType.Extent.ATTRIBUTE, 10)])`

Reply:
I thought about this and concluded that it's never valid for a lower limit to be of `ArrayType.Extent` since it can only be this if the upper limit is and in that case there should just be a single `ArrayType.Extent.xxx` rather than a tuple of values. I've simplified the checking code.